File: onchain-verifier/scripts/verifier.py
"""
On-Chain Verifier - Python 包装器

用于 SpoonOS Agent 集成的 Python 接口
通过子进程调用 TypeScript 实现
"""

import os
import json
import subprocess
from typing import Dict, List, Any, Optional
from datetime import datetime
from dataclasses import dataclass


@dataclass
class TransferEvent:
    """转账事件"""
    from_address: str
    to_address: str
    token_id: int
    block_number: int
    transaction_hash: str
    timestamp: int
    event_type: str  # 'mint' | 'transfer' | 'burn'


@dataclass
class AirdropVerification:
    """空投验证结果"""
    total_amount: int
    unique_recipients: int
    transactions: List[TransferEvent]
    fulfillment_rate: float
    evidence_strength: str  # 'strong' | 'moderate' | 'weak' | 'none'
    summary: str


class OnChainVerifier:
    """链上数据验证器 - Python 接口"""

    def __init__(
        self,
        rpc_url: str,
        contract_address: str,
        contract_abi: List[Dict],
        chain_id: Optional[int] = None
    ):
        """
        初始化验证器

        Args:
            rpc_url: RPC 节点 URL
            contract_address: NFT 合约地址
            contract_abi: 合约 ABI
            chain_id: 链 ID（可选）
        """
        self.rpc_url = rpc_url
        self.contract_address = contract_address
        self.contract_abi = contract_abi
        self.chain_id = chain_id

        # 验证 Node.js 环境
        self._check_nodejs()

    def query_transfer_events(
        self,
        start_time: Optional[datetime] = None,
        end_time: Optional[datetime] = None,
        start_block: Optional[int] = None,
        end_block: Optional[int] = None,
        from_address: Optional[str] = None,
        to_address: Optional[str] = None,
        event_type: str = 'all'
    ) -> List[TransferEvent]:
        """
        查询 Transfer 事件

        Args:
            start_time: 开始时间
            end_time: 结束时间
            start_block: 开始区块
            end_block: 结束区块
            from_address: 发送方地址
            to_address: 接收方地址
            event_type: 事件类型 ('mint' | 'transfer' | 'burn' | 'all')

        Returns:
            转账事件列表
        """
        params = {
            'action': 'queryTransferEvents',
            'config': self._get_config(),
            'params': {
                'startTime': start_time.isoformat() if start_time else None,
                'endTime': end_time.isoformat() if end_time else None,
                'startBlock': start_block,
                'endBlock': end_block,
                'from': from_address,
                'to': to_address,
                'eventType': event_type,
            }
        }

        result = self._call_typescript(params)
        return [self._parse_transfer_event(e) for e in result]

    def verify_airdrop_promise(
        self,
        promise: Dict[str, Any],
        project_address: str,
        start_time: datetime,
        end_time: datetime,
        min_batch_size: int = 5
    ) -> AirdropVerification:
        """
        验证空投承诺

        Args:
            promise: 承诺对象
            project_address: 项目方地址
            start_time: 开始时间
            end_time: 结束时间
            min_batch_size: 最小批量大小

        Returns:
            空投验证结果
        """
        params = {
            'action': 'verifyAirdropPromise',
            'config': self._get_config(),
            'params': {
                'promise': promise,
                'projectAddress': project_address,
                'startTime': start_time.isoformat(),
                'endTime': end_time.isoformat(),
                'minBatchSize': min_batch_size,
            }
        }

        result = self._call_typescript(params)
        return self._parse_airdrop_verification(result)

    def format_for_llm(
        self,
        promise: Dict[str, Any],
        onchain_data: AirdropVerification,
        mode: str = 'full'
    ) -> str:
        """
        格式化为 LLM 上下文

        Args:
            promise: 承诺对象
            onchain_data: 链上数据
            mode: 格式模式 ('full' | 'compact' | 'scoring')

        Returns:
            格式化的上下文字符串
        """
        params = {
            'action': 'formatForLLM',
            'config': self._get_config(),
            'params': {
                'promise': promise,
                'onchainData': {
                    'totalAmount': onchain_data.total_amount,
                    'uniqueRecipients': onchain_data.unique_recipients,
                    'fulfillmentRate': onchain_data.fulfillment_rate,
                    'evidenceStrength': onchain_data.evidence_strength,
                    'summary': onchain_data.summary,
                },
                'mode': mode,
            }
        }

        return self._call_typescript(params)

    # ==================== 私有方法 ====================

    def _get_config(self) -> Dict[str, Any]:
        """获取配置"""
        return {
            'rpcUrl': self.rpc_url,
            'contractAddress': self.contract_address,
            'contractABI': self.contract_abi,
            'chainId': self.chain_id,
        }

    def _call_typescript(self, params: Dict[str, Any]) -> Any:
        """
        调用 TypeScript 实现

        Args:
            params: 参数

        Returns:
            执行结果
        """
        # 获取脚本路径
        script_dir = os.path.dirname(os.path.abspath(__file__))
        ts_script = os.path.join(script_dir, 'verifier_cli.ts')

        # 调用 ts-node
        try:
            result = subprocess.run(
                ['ts-node', ts_script],
                input=json.dumps(params),
                capture_output=True,
                text=True,
                check=True
            )

            return json.loads(result.stdout)

        except subprocess.CalledProcessError as e:
            raise Exception(f"TypeScript execution failed: {e.stderr}")

        except json.JSONDecodeError as e:
            raise Exception(f"Failed to parse TypeScript output: {e}")

    def _check_nodejs(self):
        """检查 Node.js 环境"""
        try:
            subprocess.run(
                ['node', '--version'],
                capture_output=True,
                check=True
            )
        except (subprocess.CalledProcessError, FileNotFoundError):
            raise Exception(
                "Node.js not found. Please install Node.js 18+ to use this tool."
            )

        try:
            subprocess.run(
                ['ts-node', '--version'],
                capture_output=True,
                check=True
            )
        except (subprocess.CalledProcessError, FileNotFoundError):
            raise Exception(
                "ts-node not found. Please install: npm install -g ts-node typescript"
            )

    def _parse_transfer_event(self, data: Dict[str, Any]) -> TransferEvent:
        """解析转账事件"""
        return TransferEvent(
            from_address=data['from'],
            to_address=data['to'],
            token_id=int(data['tokenId']),
            block_number=data['blockNumber'],
            transaction_hash=data['transactionHash'],
            timestamp=data['timestamp'],
            event_type=data['eventType']
        )

    def _parse_airdrop_verification(
        self,
        data: Dict[str, Any]
    ) -> AirdropVerification:
        """解析空投验证结果"""
        transactions = [
            self._parse_transfer_event(t)
            for t in data.get('transactions', [])
        ]

        return AirdropVerification(
            total_amount=data['totalAmount'],
            unique_recipients=data['uniqueRecipients'],
            transactions=transactions,
            fulfillment_rate=data['fulfillmentRate'],
            evidence_strength=data['evidenceStrength'],
            summary=data['summary']
        )


# ==================== SpoonOS 工具集成 ====================

from spoonos import BaseTool


class OnChainQueryTool(BaseTool):
    """链上查询工具 - SpoonOS 集成"""

    name: str = "onchain_query"
    description: str = "查询链上数据验证项目承诺"

    def __init__(self):
        super().__init__()
        self.verifier = OnChainVerifier(
            rpc_url=os.getenv('RPC_URL'),
            contract_address=os.getenv('NFT_CONTRACT_ADDRESS'),
            contract_abi=self._load_abi()
        )

    async def execute(
        self,
        promise: Dict[str, Any],
        project_address: str,
        start_date: str,
        end_date: str
    ) -> Dict[str, Any]:
        """
        执行链上查询

        Args:
            promise: 承诺对象
            project_address: 项目方地址
            start_date: 开始日期 (ISO 格式)
            end_date: 结束日期 (ISO 格式)

        Returns:
            验证结果
        """
        from datetime import datetime

        result = self.verifier.verify_airdrop_promise(
            promise=promise,
            project_address=project_address,
            start_time=datetime.fromisoformat(start_date),
            end_time=datetime.fromisoformat(end_date)
        )

        return {
            "found": result.total_amount > 0,
            "evidence": {
                "total_amount": result.total_amount,
                "unique_recipients": result.unique_recipients,
                "fulfillment_rate": result.fulfillment_rate,
                "evidence_strength": result.evidence_strength,
                "summary": result.summary
            },
            "fulfillment_rate": result.fulfillment_rate
        }

    def _load_abi(self) -> List[Dict]:
        """加载合约 ABI"""
        # 从 DDC Market SDK 加载 ABI
        abi_path = os.path.join(
            os.path.dirname(__file__),
            '../node_modules/@ddcmarket/sdk/dist/esm/abi/DDCNFT.json'
        )

        if os.path.exists(abi_path):
            with open(abi_path, 'r') as f:
                return json.load(f)

        # 如果找不到，使用最小 ABI
        return [
            {
                "anonymous": False,
                "inputs": [
                    {"indexed": True, "name": "from", "type": "address"},
                    {"indexed": True, "name": "to", "type": "address"},
                    {"indexed": True, "name": "tokenId", "type": "uint256"}
                ],
                "name": "Transfer",
                "type": "event"
            }
        ]


# ==================== 测试代码 ====================

if __name__ == "__main__":
    print("Testing On-Chain Verifier (Python Wrapper)...\n")

    # 测试配置
    verifier = OnChainVerifier(
        rpc_url=os.getenv('RPC_URL', 'https://eth-mainnet.g.alchemy.com/v2/demo'),
        contract_address='0x...',  # 替换为实际合约地址
        contract_abi=[]  # 使用最小 ABI
    )

    print("✓ Verifier initialized")
    print("\nTo test with real data:")
    print("1. Set RPC_URL environment variable")
    print("2. Set NFT_CONTRACT_ADDRESS environment variable")
    print("3. Run: python verifier.py")

File: onchain-verifier/scripts/verifier.ts
/**
 * On-Chain Verifier - 核心实现
 *
 * 基于 ethers.js v6 的链上数据查询和验证工具
 * 用于 Promise Breaker Detector 项目
 */

import { ethers, Contract, Provider, JsonRpcProvider } from 'ethers';

export interface VerifierConfig {
  rpcUrl: string;
  contractAddress: string;
  contractABI: any[];
  chainId?: number;
  fallbackRpcUrl?: string;
  timeout?: number;
}

export interface TransferEvent {
  from: string;
  to: string;
  tokenId: bigint;
  blockNumber: number;
  transactionHash: string;
  timestamp: number;
  eventType: 'mint' | 'transfer' | 'burn';
}

export interface AirdropVerification {
  totalAmount: number;
  uniqueRecipients: number;
  transactions: TransferEvent[];
  fulfillmentRate: number;
  evidenceStrength: 'strong' | 'moderate' | 'weak' | 'none';
  summary: string;
}

export class OnChainVerifier {
  private provider: JsonRpcProvider;
  private contract: Contract;
  private config: VerifierConfig;

  constructor(config: VerifierConfig) {
    this.config = config;
    this.provider = new ethers.JsonRpcProvider(config.rpcUrl);
    this.contract = new Contract(
      config.contractAddress,
      config.contractABI,
      this.provider
    );
  }

  /**
   * 查询 Transfer 事件
   */
  async queryTransferEvents(params: {
    startTime?: Date;
    endTime?: Date;
    startBlock?: number;
    endBlock?: number;
    from?: string;
    to?: string;
    eventType?: 'mint' | 'transfer' | 'burn' | 'all';
  }): Promise<TransferEvent[]> {
    // 1. 转换时间为区块号
    let startBlock = params.startBlock;
    let endBlock = params.endBlock;

    if (params.startTime && !startBlock) {
      startBlock = await this.getBlockByTimestamp(params.startTime);
    }

    if (params.endTime && !endBlock) {
      endBlock = await this.getBlockByTimestamp(params.endTime);
    }

    // 2. 构建事件过滤器
    const filter = this.buildTransferFilter(params);

    // 3. 查询事件
    const events = await this.contract.queryFilter(
      filter,
      startBlock,
      endBlock
    );

    // 4. 解析事件
    const transfers: TransferEvent[] = [];
    for (const event of events) {
      const transfer = await this.parseTransferEvent(event);
      transfers.push(transfer);
    }

    return transfers;
  }

  /**
   * 验证空投承诺
   */
  async verifyAirdropPromise(params: {
    promise: any;
    projectAddress: string;
    startTime: Date;
    endTime: Date;
    minBatchSize?: number;
  }): Promise<AirdropVerification> {
    const minBatchSize = params.minBatchSize || 5;

    // 1. 查询从项目方地址发出的转账
    const transfers = await this.queryTransferEvents({
      startTime: params.startTime,
      endTime: params.endTime,
      from: params.projectAddress,
      eventType: 'transfer',
    });

    // 2. 识别空投交易（批量转账）
    const airdropTxs = this.identifyAirdropTransactions(
      transfers,
      minBatchSize
    );

    // 3. 统计数据
    const uniqueRecipients = new Set(airdropTxs.map(t => t.to)).size;
    const totalAmount = airdropTxs.length;

    // 4. 计算履约率
    const expectedAmount = params.promise.expectedAmount || 0;
    const fulfillmentRate = expectedAmount > 0
      ? (totalAmount / expectedAmount) * 100
      : 0;

    // 5. 评估证据强度
    const evidenceStrength = this.assessEvidenceStrength(
      totalAmount,
      uniqueRecipients,
      fulfillmentRate
    );

    // 6. 生成摘要
    const summary = `Found ${totalAmount} airdrop transaction(s) to ${uniqueRecipients} unique recipient(s)`;

    return {
      totalAmount,
      uniqueRecipients,
      transactions: airdropTxs,
      fulfillmentRate,
      evidenceStrength,
      summary,
    };
  }

  /**
   * 获取合约统计信息
   */
  async getContractStatistics(): Promise<{
    totalSupply: number;
    totalMinted: number;
    totalBurned: number;
    uniqueHolders: number;
  }> {
    // 实现统计逻辑
    // TODO: 实现完整的统计功能
    throw new Error('Not implemented yet');
  }

  /**
   * 格式化为 LLM 上下文
   */
  async formatForLLM(params: {
    promise: any;
    onchainData: AirdropVerification;
    mode: 'full' | 'compact' | 'scoring';
  }): Promise<string> {
    if (params.mode === 'compact') {
      return this.formatCompact(params.promise, params.onchainData);
    } else if (params.mode === 'scoring') {
      return this.formatForScoring(params.promise, params.onchainData);
    } else {
      return this.formatFull(params.promise, params.onchainData);
    }
  }

  // ==================== 私有方法 ====================

  private async getBlockByTimestamp(timestamp: Date): Promise<number> {
    // 二分查找最接近的区块
    // TODO: 实现二分查找算法
    const currentBlock = await this.provider.getBlockNumber();
    return currentBlock - 1000; // 临时实现
  }

  private buildTransferFilter(params: any) {
    const { from, to, eventType } = params;

    if (eventType === 'mint') {
      return this.contract.filters.Transfer(ethers.ZeroAddress, to, null);
    } else if (eventType === 'burn') {
      return this.contract.filters.Transfer(from, ethers.ZeroAddress, null);
    } else {
      return this.contract.filters.Transfer(from, to, null);
    }
  }

  private async parseTransferEvent(event: any): Promise<TransferEvent> {
    const block = await this.provider.getBlock(event.blockNumber);

    let eventType: 'mint' | 'transfer' | 'burn';
    if (event.args.from === ethers.ZeroAddress) {
      eventType = 'mint';
    } else if (event.args.to === ethers.ZeroAddress) {
      eventType = 'burn';
    } else {
      eventType = 'transfer';
    }

    return {
      from: event.args.from,
      to: event.args.to,
      tokenId: event.args.tokenId,
      blockNumber: event.blockNumber,
      transactionHash: event.transactionHash,
      timestamp: block!.timestamp,
      eventType,
    };
  }

  private identifyAirdropTransactions(
    transfers: TransferEvent[],
    minBatchSize: number
  ): TransferEvent[] {
    // 按交易哈希分组
    const txGroups = new Map<string, TransferEvent[]>();

    for (const transfer of transfers) {
      const txHash = transfer.transactionHash;
      if (!txGroups.has(txHash)) {
        txGroups.set(txHash, []);
      }
      txGroups.get(txHash)!.push(transfer);
    }

    // 筛选批量交易
    const airdropTxs: TransferEvent[] = [];
    for (const [_, group] of txGroups) {
      if (group.length >= minBatchSize) {
        airdropTxs.push(...group);
      }
    }

    return airdropTxs;
  }

  private assessEvidenceStrength(
    totalAmount: number,
    uniqueRecipients: number,
    fulfillmentRate: number
  ): 'strong' | 'moderate' | 'weak' | 'none' {
    if (totalAmount === 0) return 'none';
    if (fulfillmentRate >= 80 && uniqueRecipients >= 100) return 'strong';
    if (fulfillmentRate >= 50 && uniqueRecipients >= 50) return 'moderate';
    return 'weak';
  }

  private formatFull(promise: any, data: AirdropVerification): string {
    return `
=== PROMISE ===
Content: ${promise.content}
Deadline: ${promise.deadline}

=== ON-CHAIN EVIDENCE ===
Transactions Found: ${data.totalAmount}
Unique Recipients: ${data.uniqueRecipients}
Fulfillment Rate: ${data.fulfillmentRate.toFixed(1)}%
Evidence Strength: ${data.evidenceStrength}

Summary: ${data.summary}

=== PRELIMINARY ANALYSIS ===
${this.generateAnalysis(promise, data)}
    `.trim();
  }

  private formatCompact(promise: any, data: AirdropVerification): string {
    return `Promise: ${promise.content}\nEvidence: ${data.totalAmount} txs, ${data.uniqueRecipients} recipients, ${data.fulfillmentRate.toFixed(0)}% fulfilled`;
  }

  private formatForScoring(promise: any, data: AirdropVerification): string {
    return `
=== SCORING CONTEXT ===
Promise: ${promise.content}
Evidence: ${data.totalAmount} transactions, ${data.uniqueRecipients} recipients
Fulfillment: ${data.fulfillmentRate.toFixed(1)}%
Strength: ${data.evidenceStrength}
    `.trim();
  }

  private generateAnalysis(promise: any, data: AirdropVerification): string {
    const timing = data.fulfillmentRate > 0 ? 'Completed' : 'Not completed';
    const coverage = `${data.uniqueRecipients} recipients`;
    const overall = data.evidenceStrength === 'strong'
      ? 'Strong evidence - likely fulfilled'
      : data.evidenceStrength === 'moderate'
      ? 'Moderate evidence - partially fulfilled'
      : 'Weak evidence - likely not fulfilled';

    return `Timing: ${timing}\nCoverage: ${coverage}\nOverall: ${overall}`;
  }
}

// ==================== 导出 ====================

export default OnChainVerifier;
